### PR TITLE
88 create a create course iteration view

### PR DESCRIPTION
**Pull request by jaess105**
# Task

Create a view in which course-iterations can be created.

## Functionality

* GET
  * When there are no courses to which a course-iteration can belong, an error is displayed.
  * Otherwise, a form is displayed in which the user can select values for each field.
    * A submit button is displayed to send the data back to the server
    * The get request can optionally be called with parameters, which are displayed as errors at their corresponding field, if there is a corresponding field. If there is none, the parameters are ignored.
* POST
  * Checks if the sent data is valid.
    * If it is valid, a course-iteration is saved in the database and a success message is displayed.
    * Otherwise, a URI is constructed adding the errors as parameters to a request which should be sent back to the original GET request. The destination is injected, so it should point back to the URI of the get.

## Changes

* Added course_iteration_creation_controller.clj containing the GET and the POST mapping described above
  * Added tests
  * Added doc strings and fdefs for public methods 
* Added course_iteration_creation_view.clj containing the different functions creating views for the get and post with hiccup.
  * Added tests
  * Added doc strings and fdefs for public methods
* Added the GET and POST methods to the private routes of our app. 
* Added services for course-iteration, course, and question-set
  * provided specs for all of them
  * They are currently mostly wrapping db functions. When there are repositories, these should be used instead. Added todo for that. 

## Other changes
* Added .clj-kondo/config.edn in which unresolved vars can be excluded from the static analysis. The ignored vars are generated dynamically, which is why they are not found by kondo and must be ignored to pass the kondo check.
* Moved html-response function into `util.ring-extensions` as it was needed in another module and would lead to a circular import in main. (Should not lead to merge conflicts, unless someone is using it in another module as well.)
* Refactored: `get-all-courses` was fetching all course-iterations. Added a function to fetch all course-iterations and fixed `get-all-courses`
* Updated Specs
  * Added a spec ` :general/non-blank-string `
    * Used that spec for some already used specs
  * Specified `:course-iteration/semester` spec
  * Added macro for specs to generate a spec for a map without namespaced keys

### CODE REVIEW

File: src/clj/core.clj
Comment (matlorr) [inline src/clj/core.clj:1]: Looks neat to me :), I dont yet understand though why in line 40 and 41 a call to partial is necessary when providing service functions to the routes.

File: test/controllers/course_iteration/course_iteration_controller_test.clj
Comment (matlorr) [inline test/controllers/course_iteration/course_iteration_controller_test.clj:1]: I cant find a assertion for the test in lines 61 - 87.

Comment (jaess105) [reply test/controllers/course_iteration/course_iteration_controller_test.clj:1]: For the Test from line 62, the assertions are in the `db-add-fun-stub` (68-72). Here the function is asserting that the passed in parameters are correct. And for the test in line 89 there is the assertion inside of `t/are` (109-114). `t/are` is a form of parameterized testing.

File: src/clj/core.clj
Comment (jaess105) [reply src/clj/core.clj:1]: Even though we just discussed it in the meeting: The partial is used to pass in a function from the service while the underlying implementation of the controller is unaware of the actual service. This way any function can be passed into the controller and we can use the implementation of the service.

**Review by matlorr — APPROVED** [review]
All open question were answered and the code looks good to me!

**Pull request merged**